Enable reference counting for multiple nested open sessions

Krebrov001 opened the pull request:
Enable the PlatformService to support clients having multiple
independent components, which can concurrently use the single
connection. Establish a shared pointer-esque abstraction over
connections, a one to one mapping with the client process.

Add the reference count to the JSON active session schema.

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

- Relates to #2011 
- Fixes #2011

Krebrov001 committed:
Enable reference counting for multiple nested open sessions

Enable the PlatformService to support clients having multiple
independent components, which can concurrently use the simgle
connection. Establish a shared pointer esque abstraction over
connections, a one to one mapping with the client process.

Add the reference count to the JSON active session schema.

Signed-off-by: Konstantin Rebrov <konstantin.rebrov@intel.com>

cmcantalupo committed:
Add check for docstring_parser when generating xml file

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo commented on service/geopmdpy/system_files.py:
```diff
@@ -570,6 +572,91 @@ def set_watch_id(self, client_pid, watch_id):
+            client_pid (int): Linux PID that opened the session
+
+        Returns:
+            int: GLib watch ID to track the session lifetime
```
```suggestion
            int: The reference count
```

cmcantalupo commented on service/geopmdpy/system_files.py:
```diff
@@ -570,6 +572,91 @@ def set_watch_id(self, client_pid, watch_id):
+        self.check_client_active(client_pid, 'get_reference_count')
+        return self._sessions[client_pid]['reference_count']
+
+    def set_reference_count(self, client_pid, reference_count):
```
This interface only seems to be used in the test.  We can leave it, but I'm not sure how it would be used in practice.

cmcantalupo committed:
Update documentation strings in io.github.geopm.xml

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Update io.github.geopm.xml documentation

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Update documentation for reference count

- Simplify some details.
- Make sure documentation is applicable to DBus interface.
- Get rid of an early return.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fixup return value documentation (copy/paste error)

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

bgeltz commented on service/geopmdpy/service.py:
```diff
@@ -349,11 +358,15 @@ def open_session(self, user, client_pid):
                               the session.
 
         """
+        # if the connection already exists
```
```suggestion
```

bgeltz commented on service/geopmdpy/service.py:
```diff
@@ -349,11 +358,15 @@ def open_session(self, user, client_pid):
+            self._active_sessions.increment_reference_count(client_pid)
+        else:
+            # Establish the connection and add an active session with
+            # the initial reference count of 1.
```
```suggestion
```

bgeltz commented on service/geopmdpy/service.py:
```diff
@@ -370,10 +383,64 @@ def close_session(self, client_pid):
+        self._active_sessions.check_client_active(client_pid, 'PlatformCloseSession')
+        reference_count = self._active_sessions.get_reference_count(client_pid)
+        # A negative reference_count is impossible because this condition
+        # is enforced by the _session_schema.
```
```suggestion
```

bgeltz commented on service/geopmdpy/service.py:
```diff
@@ -370,10 +383,64 @@ def close_session(self, client_pid):
+            # that in the process of _close_Session_completely so we
+            # need to call _close_Session_completely() again. The
+            # daemon died, was restarted, and found a session with
+            # zero reference count.
```
```suggestion
            # The daemon died, restarted, and found a session with
            # zero reference count.
```

bgeltz commented on service/geopmdpy/service.py:
```diff
@@ -370,10 +383,64 @@ def close_session(self, client_pid):
+            # zero reference count.
+            self._close_session_completely(client_pid)
+        elif reference_count == 1:
+            # last reference, close the session completely
```
```suggestion
```

bgeltz commented on service/geopmdpy/service.py:
```diff
@@ -370,10 +383,64 @@ def close_session(self, client_pid):
+            self._active_sessions.decrement_reference_count(client_pid)
+            self._close_session_completely(client_pid)
+        else:  # reference_count > 1:
+            # more than one reference, decrement the reference count
```
```suggestion
```

bgeltz reviewed (changes requested):
Do we need to add a destructor to the `PlatformService` class that calls `close_session_completely()` for all the client PIDs that are active?  I expected to see this chunk of code somewhere.

cmcantalupo committed:
Remove excessive comments in code.

- These comments are not incorrect, but
  more detail than is nessesary.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>
Co-authored-by: Brad Geltz <brgeltz@gmail.com>
